starlark/src/eval/def.rs

```diff
@@ -71,54 +78,186 @@ impl Def {
+                            )
+                        })
+                        .collect(),
+                ),
+                Expr::Identifier(ident) => match locals.get(&ident.node) {
+                    Some(&slot) => Expr::Slot(slot, ident),
+                    None => Expr::Identifier(ident),
+                },
+                Expr::Slot(..) => unreachable!(),
+                Expr::Dot(left, right) => {
+                    Expr::Dot(Def::transform_locals_in_expr(left, locals), right)
+                }
+                Expr::Call(function, args, kwargs, star_args, star_star_kwargs) => Expr::Call(
+                    Def::transform_locals_in_expr(function, locals),
+                    args.into_iter()
+                        .map(|expr| Def::transform_locals_in_expr(expr, locals))
+                        .collect(),
+                    kwargs
+                        .into_iter()
+                        .map(|(name, value)| (name, Def::transform_locals_in_expr(value, locals)))
+                        .collect(),
+                    star_args.map(|expr| Def::transform_locals_in_expr(expr, locals)),
+                    star_star_kwargs.map(|expr| Def::transform_locals_in_expr(expr, locals)),
+                ),
+                Expr::ArrayIndirection(array, index) => Expr::ArrayIndirection(
+                    Def::transform_locals_in_expr(array, locals),
+                    Def::transform_locals_in_expr(index, locals),
+                ),
+                Expr::Slice(array, p1, p2, p3) => Expr::Slice(
+                    array,
+                    p1.map(|expr| Def::transform_locals_in_expr(expr, locals)),
+                    p2.map(|expr| Def::transform_locals_in_expr(expr, locals)),
+                    p3.map(|expr| Def::transform_locals_in_expr(expr, locals)),
+                ),
+                Expr::Not(expr) => Expr::Not(Def::transform_locals_in_expr(expr, locals)),
+                Expr::Minus(expr) => Expr::Minus(Def::transform_locals_in_expr(expr, locals)),
+                Expr::Plus(expr) => Expr::Plus(Def::transform_locals_in_expr(expr, locals)),
+                Expr::Op(op, left, right) => Expr::Op(
+                    op,
+                    Def::transform_locals_in_expr(left, locals),
+                    Def::transform_locals_in_expr(right, locals),
+                ),
+                Expr::If(cond, then_expr, else_expr) => Expr::If(
+                    Def::transform_locals_in_expr(cond, locals),
+                    Def::transform_locals_in_expr(then_expr, locals),
+                    Def::transform_locals_in_expr(else_expr, locals),
+                ),
+                n @ Expr::IntLiteral(..) | n @ Expr::StringLiteral(..) => n,
+                n @ Expr::DictComprehension(..)
+                | n @ Expr::ListComprehension(..)
+                | n @ Expr::SetComprehension(..) => {
+                    // TODO: be smarter
```

Comment:
Can you be more precise? i.e. what is suboptimal and we should optimize here?

Reply:
Updated (locally) with this comment:

```
                    // TODO: do the same access-by-index optimization for compr locals.
                    // TODO: access parent scope (function) variables by index, not by name.
```